include/field.h

```diff
@@ -433,6 +435,26 @@ struct filter {
                                            const Store* store,
                                            const std::string& doc_id_prefix,
                                            std::vector<filter>& filters);
+
+    static Option<bool> parse_filter_query2(const std::string& filter_query,
+                                            const std::unordered_map<std::string, field>& search_schema,
+                                            const Store* store,
+                                            const std::string& doc_id_prefix,
+                                            filter_node_t* root);
+};
+
+struct filter_node_t {
+    std::string token;
+    filter_node_t *left;
+    filter_node_t *right;
+    filter filter;
```

Comment:
The build is not compiling because of naming the variable as the same as the type: 

```
error: declaration of 'filter filter_node_t::filter' changes meaning of 'filter'
```

We can call it as `filter_exp`.